### PR TITLE
vfio: fix the bug associated with the boot sequence

### PR DESCRIPTION
**Pull request by PandaWei**
the case vfio.primary_boot.nic suffer from the following,
1) VM starts normally, while alter the boot sequence and
   set NIC PCI device as the only and first boot option.
   It should report no bootable device and startup fails.
2) The OS boot element and PCI Device boot elements are
   mutually exclusive.
   It's necessary that remove all of OS boots and
   add PCI boot attribute before re-defining VM.
   otherwise,
   it will keep the primary configuration or throw errors,
   such as,
   'unsupported configuration: per-device boot elements
   cannot be used together with os/boot elements'

This patch can fix all of these problems.

### CODE REVIEW

**Comment by PandaWei** [comment]
It needs virttest#2082

**Comment by PandaWei** [comment]
[root@PQ1000 virt]# ./run -t libvirt --no-downloads -k --keep-image-between-tests --test "type_specific.io-github-autotest-libvirt.vfio.primary_boot.nic"
SETUP: PASS (0.40 s)
DATA DIR: /home/autotest_fjtest_v1.5.4/client/tests/virt/shared/data
DEBUG LOG: /home/autotest_fjtest_v1.5.4/client/tests/virt/logs/run-2015-03-19-16.48.37/debug.log
TESTS: 1
(1/1) type_specific.io-github-autotest-libvirt.vfio.primary_boot.nic: PASS (143.19 s)
TOTAL TIME: 143.25 s (02:23)
TESTS PASSED: 1
TESTS FAILED: 0
SUCCESS RATE: 100.00 %

**Comment by PandaWei** [comment]
Add new commit
vfio: make some valuable adjustments
        * some cases involves formatting disk and wipe all of data,
      that stored at Fibre PCI device, which is dangerous and
      unfriendly for user.
      This commit allows to choose whether perform these steps.
    * optimize steps for attaching PCI device.

**Comment by PandaWei** [comment]
./run -t libvirt --no-downloads -k --keep-image-between-tests --test "type_specific.io-github-autotest-libvirt.vfio" 
SETUP: PASS (0.40 s)
DATA DIR: /home/autotest_fjtest_v1.5.4/client/tests/virt/shared/data
DEBUG LOG: /home/autotest_fjtest_v1.5.4/client/tests/virt/logs/run-2015-03-24-16.29.57/debug.log
TESTS: 6
(1/6) type_specific.io-github-autotest-libvirt.vfio.nic_group: PASS (92.19 s)
(2/6) type_specific.io-github-autotest-libvirt.vfio.fibre_group: PASS (145.79 s)
(3/6) type_specific.io-github-autotest-libvirt.vfio.nic_and_fibre_group: PASS (168.20 s)
(4/6) type_specific.io-github-autotest-libvirt.vfio.nic_single_to_iommu_group:  PASS (35.60 s)
(5/6) type_specific.io-github-autotest-libvirt.vfio.primary_boot.nic: PASS (144.50 s)
(6/6) type_specific.io-github-autotest-libvirt.vfio.primary_boot.fibre: PASS (146.31 s)
TOTAL TIME: 732.69 s (12:12)
TESTS PASSED: 6
TESTS FAILED: 0
SUCCESS RATE: 100.00 %

File: libvirt/tests/src/vfio.py [hunk 1]
```diff
@@ -309,10 +299,9 @@ def test_nic_group(vm, params):
             raise error.TestFail("Login vm successfully, but not expected.")
         except remote.LoginTimeoutError:
             logging.debug("Expected failure.")
-            return
-        finally:
             vm.destroy(gracefully=False)
             cleanup_devices(pci_id, device_type)
+            return
```
Comment (will-Do) [inline libvirt/tests/src/vfio.py]: Is this ```return``` necessary ?

Comment (PandaWei) [inline libvirt/tests/src/vfio.py]: @will-Do  thanks .
Yes, it's necessary
first,
I just move ‘return’  to the back of  the cleanup steps, such as  vm.destroy and cleanup.
It ensures that  completely clean up the environment.
secondly,
it's a negative case that we set the PCI as the first boot device.
The VM shouldn't  start normally.
If it starts up without any error,  it's abnormal, 
Maybe the configuration of test is wrong,
so,  it's no point to run the next step.

**Comment by Hao-Liu** [comment]
LGTM. Thanks @PandaWei @will-Do. Merging this.

**Pull request merged**